Debugger MemoryWidget: More display types, use combo box for options

TryTwo opened the pull request:
Added Int8/16/32 and double..

![PR_display](https://user-images.githubusercontent.com/10532806/161950146-f4d468e3-57e9-4e3e-af84-a1af246689a5.jpg)

- Do we want signed displays?
- 1.234e+09 syntax looks like it could be better, but what would I replace it with? Remove the +?
- Int8 looks terrible, but will look fine if we add a dual view  - two different column types with 4 bytes per row
- I left display type as a group, as it will likely get more items soon.
- Do we need to rename U8/16/32?
- Thoughts on sidebar rearrangement?

Most importantly, what else do I need to add? Do we need more customization in this PR like bytes per row?  A dual-view would require a lot of rearranging in the Update function to pass two view types through it. I think I would spin off column value updates and breakpoint tagging updates to two different functions, but will take suggestions on a better method to do it.

TryTwo commented:
Added row length combo for 4 bytes, 8 bytes, or 16 bytes. Helps a lot with integer displays.

AdmiralCurtiss commented:
The Search buttons should probably be closer to the top, probably right below the Set Value button, since they work of the same input field. Also get rid of the box they're in.

The display type might also look better with just a label next to it instead of an entire box for a single button.

AdmiralCurtiss commented:
Rename U8/16/32 to Hex, probably. Might as well have Signed N and Unsigned N as separate options too.

Integers should be right-aligned within the column.

The Int32 view needs to be wide enough to fit the full integer, *please* don't use exponential for that, it's completely useless.

TryTwo commented:
![PR_display2](https://user-images.githubusercontent.com/10532806/162105340-09bdd29e-f28e-4b43-9eba-351aad4b015a.jpg)

AdmiralCurtiss commented:
...you should rebase this on master before you introduce even more conflicts.

TryTwo commented:
Oh I didn't know it'd pick up rebase changes in Compare. That's a bit annoying.

TryTwo commented:
I don't know how far this PR should go. I can break things up into two PRs, which is why I'm making multiple commits. Should I add this to this PR?
![PR_display3](https://user-images.githubusercontent.com/10532806/162129335-332090bc-d4de-4395-a5d4-1b36507c0b05.jpg)

AdmiralCurtiss commented:
Less changes per PR is better as long as they make sense individually -- easier to review and all. So yes, I'd split that up.

It's okay to have multiple PRs that depend on eachother, too, so you don't have to deal with merge conflicts. That is, one PR that has commit A only, and then another PR that has commits A and B with a note that the first PR should be merged before this one.

TryTwo commented:
Ahh ok, that makes sense.

TryTwo commented:
Made clicking preserve byte offset (like 80000001), made row breakpoint conform to row settings.

Breakpointing odd bytes (0x1) has weird results, I'm not sure if I need to enforce some sort of even byte check. Also, a 4-byte row breakpoint on a 0x2 offset is weird.

TryTwo commented:
@AdmiralCurtiss  I need your thoughts on how the rows should be aligned if you input 80000001 or 80000002.  Keep in mind integer values change if the alignment isn't a multiple of the underlying byte length.  The easy thing to do is forcibly round the row addresses to 0 at all times and just highlight the target address so you know where to look, but I'm thinking you need more functionality than that.

4 byte and 8 byte row lengths would of course if you rows aligned like that: Like 80000000, 80000004, 80000008 for byte.

sepalani commented:
> @AdmiralCurtiss I need your thoughts on how the rows should be aligned if you input 80000001 or 80000002. Keep in mind integer values change if the alignment isn't a multiple of the underlying byte length. The easy thing to do is forcibly round the row addresses to 0 at all times and just highlight the target address so you know where to look, but I'm thinking you need more functionality than that.
> 
> 4 byte and 8 byte row lengths would of course if you rows aligned like that: Like 80000000, 80000004, 80000008 for byte.

What about adding a setting for that and let the user set the alignment such as: _(feel free to come with better names/wording)_
 - auto: alignment based on type and the number of bytes per row
 - none: will go at the address+offset without taking the alignment into account
 - fixed: based on the number of bytes per row (the behaviour of hex editor)
 - custom: based on the user input

I'm fine with moving the memory dump buttons down below as long as they are in a dedicated group box like the other options.

sepalani commented on Source/Core/DolphinQt/Debugger/MemoryViewWidget.cpp:
```diff
@@ -63,28 +63,28 @@ void MemoryViewWidget::UpdateFont()
 }
 
-static int GetColumnCount(MemoryViewWidget::Type type)
+static int GetByteLength(MemoryViewWidget::Type type)
```
A `byte` or rather an [octet](https://en.wikipedia.org/wiki/Byte) is always 8 bits so "byte length" doesn't really make sense in this context. A better wording would be type length or type size, I guess.

TryTwo replied:
ByteCount? I think it needs the word byte in there to make the value returned obvious, as bits are referenced often..

sepalani replied:
Often in C/C++ you'll find `size`/`length`/`sizeof` referring to the size in bytes. Wording is hard and I don't I have a better name on top of my head right now...

TryTwo replied:
Oops, I forgot to do this, I'll include it in the next update.

TryTwo commented:
> What about adding a setting for that and let the user set the alignment such as: _(feel free to come with better names/wording)_

That might work. Would it be better to put that in the table's context menu?

AdmiralCurtiss commented:
I think by default just aligning starting at the current address is fine, if you jump to ...02 then you'll just get 02, 06, 0a, 0e when 4-byte aligned. If you want to be fancy you can make options, but I dunno if it's really all that necessary -- you can just go back 2 bytes after all.

TryTwo commented:
Alright, how is that for options?

Too many things were being called address / addr, so I renamed some things, and then I renamed the associated things as well.

AdmiralCurtiss commented:
@dolphin-emu-bot rebuild

AdmiralCurtiss commented on Source/Core/DolphinQt/Debugger/MemoryViewWidget.cpp:
```diff
@@ -191,18 +251,45 @@ void MemoryViewWidget::Update()
+      update_values(
+          [&accessors](u32 address) { return QString::number(accessors->ReadU32(address)); });
+      break;
+    case Type::Signed8:
+      update_values([&accessors](u32 address) {
+        return QString::number(Common::BitCast<signed char>(accessors->ReadU8(address)));
+      });
+      break;
+    case Type::Signed16:
+      update_values([&accessors](u32 address) {
+        return QString::number(Common::BitCast<signed short>(accessors->ReadU16(address)));
+      });
+      break;
+    case Type::Signed32:
+      update_values([&accessors](u32 address) {
+        return QString::number(Common::BitCast<signed int>(accessors->ReadU32(address)));
+      });
+      break;
```
You can and should probably use `s8`/`s16`/`s32` here for the casts.

TryTwo replied:
Oh I didn't know you could use that. Thanks

AdmiralCurtiss commented on Source/Core/DolphinQt/Debugger/MemoryViewWidget.cpp:
```diff
@@ -35,6 +36,19 @@ MemoryViewWidget::MemoryViewWidget(QWidget* parent) : QTableWidget(parent)
   setContextMenuPolicy(Qt::CustomContextMenu);
 
+  // Context menu actions.
+  m_align_fixed = new QAction(tr("Fixed"));  //, this, [this] { SetAddress(m_address); });
+  m_align_auto = new QAction(tr("Auto"));    //, this, [this] { SetAddress(m_address); });
+  m_align_none = new QAction(tr("None"));    //, this, [this] { SetAddress(m_address); });
```
What's up with this commented out code?

TryTwo replied:
Ah sorry I left a bit of trash

TryTwo commented:
The only thing left out was a custom alignment, as I didn't want to make an input box for it. If needed, I could do a "lock alignment" option, and then the address box would be used to set the alignment, then after it is locked, it wouldn't change even if the address does.

TryTwo commented:
Should be pretty much done with the PR, unless @[sepalani](https://github.com/sepalani) wants more for alignment, or if anyone has other thoughts.

For the UI buttons, search has a label that pops up giving you feedback, so I left the box alone. Then since everything else has a box, dump got one for consistency.

Dentomologist commented:
@dolphin-emu-bot rebuild

AdmiralCurtiss commented on Source/Core/DolphinQt/Debugger/MemoryWidget.cpp:
```diff
@@ -179,22 +186,32 @@ void MemoryWidget::CreateWidgets()
+  m_display_combo->addItem(tr("Hex 32"), int(MemoryViewWidget::Type::Hex32));
+  m_display_combo->addItem(tr("Unsigned 8"), int(MemoryViewWidget::Type::Unsigned8));
+  m_display_combo->addItem(tr("Unsigned 16"), int(MemoryViewWidget::Type::Unsigned16));
+  m_display_combo->addItem(tr("unsigned 32"), int(MemoryViewWidget::Type::Unsigned32));
```
This string is randomly lowercase.

AdmiralCurtiss commented on Source/Core/DolphinQt/Debugger/MemoryViewWidget.cpp:
```diff
@@ -62,28 +76,76 @@ void MemoryViewWidget::UpdateFont()
+  // Optional Alignments, always a power of 2.
+  u32 address = m_address;
+
+  if (m_align_fixed->isChecked())
+    address = address & -m_bytes_per_row;
+  else if (m_align_auto->isChecked())
+    address = address & -GetTypeSize(m_type);
```
I see what you're doing here, but some compilers are very likely to warn about taking a negative of an unsigned here. You probably want to write this as `address & ~(m_bytes_per_row - 1)` instead. Same for the GetTypeSize() one.

AdmiralCurtiss replied:
Actually, I just remembered we have a `Common::AlignDown()` function, maybe use that.

TryTwo replied:
Thanks, I was trying to find a support function

AdmiralCurtiss commented on Source/Core/DolphinQt/Debugger/MemoryViewWidget.cpp:
```diff
@@ -102,83 +164,81 @@ void MemoryViewWidget::Update()
       continue;
     }
 
-    if (m_address_space == AddressSpace::Type::Effective)
-    {
-      auto* description_item = new QTableWidgetItem(
-          QString::fromStdString(PowerPC::debug_interface.GetDescription(addr)));
-
-      description_item->setForeground(Qt::blue);
-      description_item->setFlags(Qt::ItemIsEnabled);
-
-      setItem(i, columnCount() - 1, description_item);
-    }
```
Why did you remove this?

TryTwo replied:
It  places symbol data in a column that is overwritten by normal value data.  Am I missing some case where it actually appears?

AdmiralCurtiss replied:
Mmm you're right but this is clearly trying to do something useful. Whatever, we can fix it later.

TryTwo replied:
Maybe two column mode can have address + symbol.  The symbols are what you get from codewidget.  They could also be used as bookmarks if there were additional UI stuff to make and edit them in memorywidget.

AdmiralCurtiss commented on Source/Core/DolphinQt/Debugger/MemoryWidget.cpp:
```diff
@@ -179,22 +186,32 @@ void MemoryWidget::CreateWidgets()
+  displaytype_group->setLayout(displaytype_layout);
+
+  m_display_combo = new QComboBox;
+  m_display_combo->setMaxVisibleItems(20);
```
Is this necessary?

TryTwo replied:
The combobox will have a scroll bar if we don't increase the size. I think it's better to just see all the options.

AdmiralCurtiss reviewed (commented):
I feel like the alignment option should be another combobox or something by the Display Type options, rather than hidden inside the right-click menu. The names also aren't very self-explanatory, but to be fair I don't have a good suggestion either and it's easy enough to figure out with some experimentation. Otherwise this seems mostly fine.

TryTwo commented:
Made the alignment combobox.  Also squashed the two commits to one, it was too hard to keep them split. Not sure if "No Alignment" should have a warning or be disabled for floats, doubles, and maybe ints, as the value changes.

dreamsyntax commented:
@dolphin-emu-bot rebuild

AdmiralCurtiss commented:
Alright, lemme test this again but this seems fine to me now.

AdmiralCurtiss commented:
You didn't fix the random lowercase `unsigned`! But yeah seems good otherwise.

TryTwo commented:
I got it this time.

sepalani commented:
@dolphin-emu-bot rebuild
